src/multisig/multisig_tx_builder_ringct.cpp

```diff
@@ -0,0 +1,936 @@
+        return false;
+      s = unsigned_tx.rct_signatures.p.CLSAGs[i].s;
+    }
+
+    for (std::size_t j = 0; j < ring_size; ++j) {
+      P[j] = rv.mixRing[i][j].dest;
+      C_nonzero[j] = rv.mixRing[i][j].mask;
+    }
+
+    rct::key D;
+    rct::key z;
+    auto z_wiper = epee::misc_utils::create_scope_leave_handler([&]{
+      memwipe(static_cast<rct::key *>(&z), sizeof(rct::key));
+    });
+    if (not reconstruction) {
+      sc_sub(z.bytes, sources[i].mask.bytes, a[i].bytes);  //commitment to zero privkey
+      ge_p3 H_p3;
+      rct::hash_to_p3(H_p3, rv.mixRing[i][l].dest);
+      rct::key H_l;
+      ge_p3_tobytes(H_l.bytes, &H_p3);
+      D = rct::scalarmultKey(H_l, z);  //auxilliary key image (for commitment to zero)
+      rv.p.CLSAGs[i].D = rct::scalarmultKey(D, rct::INV_EIGHT);
+      rv.p.CLSAGs[i].I = I;
+    }
+    else {
+      rv.p.CLSAGs[i].D = unsigned_tx.rct_signatures.p.CLSAGs[i].D;
+      rv.p.CLSAGs[i].I = I;
+      D = rct::scalarmultKey(rv.p.CLSAGs[i].D, rct::EIGHT);
+    }
+
+    if (not CLSAG_contexts[i].init(P, C_nonzero, C_offset, message, I, D, l, s, kAlphaComponents))
+      return false;
+
+    if (not reconstruction) {
+      rct::key mu_P;
+      rct::key mu_C;
+      if (not CLSAG_contexts[i].get_mu(mu_P, mu_C))
+        return false;
+      sc_mul(cached_w[i].bytes, mu_P.bytes, input_secret_keys[i].bytes);
+      sc_muladd(cached_w[i].bytes, mu_C.bytes, z.bytes, cached_w[i].bytes);
+    }
+  }
+  unsigned_tx.rct_signatures = std::move(rv);
+  return true;
+}
+//----------------------------------------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------------------------------------
+static bool compute_tx_fee(
+  const std::vector<cryptonote::tx_source_entry>& sources,
+  const std::vector<cryptonote::tx_destination_entry>& destinations,
+  std::uint64_t& fee
+)
+{
+  std::uint64_t in_amount = 0;
+  for (const auto& src: sources)
+    in_amount += src.amount;
+
+  std::uint64_t out_amount = 0;
+  for (const auto& dst: destinations)
+    out_amount += dst.amount;
+
+  if(out_amount > in_amount)
+    return false;
+  fee = in_amount - out_amount;
+  return true;
+}
+//----------------------------------------------------------------------------------------------------------------------
+tx_builder_ringct_t::tx_builder_ringct_t(): initialized(false) {}
+//----------------------------------------------------------------------------------------------------------------------
+tx_builder_ringct_t::~tx_builder_ringct_t()
+{
+  memwipe(static_cast<rct::key *>(cached_w.data()), cached_w.size() * sizeof(rct::key));
+}
+//----------------------------------------------------------------------------------------------------------------------
+bool tx_builder_ringct_t::init(
+  const cryptonote::account_keys& account_keys,
+  const std::vector<std::uint8_t>& extra,
+  const std::uint64_t unlock_time,
```

Comment:
While I understand why this is technically valid, the unlock time has minimal benefit and is largely considered extremely detrimental. Including it here in multisig does not seem beneficial.

Reply:
The calling code (wallet2 tx builder) expects timelocks to work as intended by the user (who has to manually specify the unlock time). Deprecating timelocks is a bigger project that's out of scope here.